screens/Scan/ScanScreen.tsx

```diff
@@ -1,28 +1,98 @@
-import React from 'react';
+import React, { useContext, useEffect, useState } from 'react';
+import { useSelector } from '@xstate/react';
+import { Linking, PermissionsAndroid, StyleSheet, View } from 'react-native';
 import { QrScanner } from '../../components/QrScanner';
-import { Column, Text } from '../../components/ui';
+import { Button, Column, Text } from '../../components/ui';
 import { Colors } from '../../components/ui/styleUtils';
 import { MainRouteProps } from '../../routes/main';
 import { MessageOverlay } from '../../components/MessageOverlay';
+import { GlobalContext } from '../../shared/GlobalContext';
 import { SendVidModal } from './SendVidModal';
 import { useScanScreen } from './ScanScreenController';
+import { selectIsActive } from '../../machines/app';
+
+const styles = StyleSheet.create({
+  buttonContainer: {
+    height: '93%',
+    width: '100%',
+  },
+  buttonStyle: {
+    position: 'absolute',
+    width: '100%',
+    bottom: -90,
+  },
+});
 
 export const ScanScreen: React.FC<MainRouteProps> = (props) => {
+  const { appService } = useContext(GlobalContext);
   const controller = useScanScreen(props);
+  const [hasPermission, setHasPermission] = useState(null);
+  const isActive = useSelector(appService, selectIsActive);
+
+  const openSettings = () => {
+    Linking.openSettings();
+  };
+
+  const locationPermission = async() => {
+    return await PermissionsAndroid.request(PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
+      {
+        title: 'Location access',
+        message: 'Location access is required for the scanning functionality.',
+        buttonNegative: 'Cancel',
+        buttonPositive: 'OK',
+      });
+  }
+
+  const enableLocation = async() => {
+    if(controller.isLocationDenied) {
+      await controller.REQUEST();
+      if(!controller.isLocationDenied) {
+        const response = await locationPermission();
+        if(response !== 'granted') {
+          openSettings()
+        }
+      }
+    } else {
+      Linking.openSettings();
+    }
+  }
+
+  useEffect(() => {
+    if (!controller.isLocationDenied && !hasPermission) {
+      (async () => {
+        const response = await locationPermission();
+        setHasPermission(response === 'granted' ? true : false);
+      })();
+    }
+  }, []);
+
+  useEffect(() => {
+    if (isActive && !controller.isLocationDenied && controller.isScanning && !hasPermission) {
+      (async () => {
+        const response = await locationPermission();
+        setHasPermission(response === 'granted' ? true : false);
+      })();
+    }
+  }, [isActive]);
```

Comment:
Can you move this into the state machine? It's causing unhandled promises because it's out of the flow. You can use as reference and include this logic into `checkingLocationService`.

We should try to keep the views lean and not include any logic.

Reply:
@pmigueld already moved, please review